### ISSUE TITLE
Add desktop's links to apply configuration on the fly

### ISSUE THREAD

**Issue by giacomocerquone** [KurtisLiggett/Simple-IP-Config#4]
As shortly explained on the title of this issue, we could implement the possibility to create links to place wherever the user want so to instantly apply the configuration. 
I worked a lot when I was 12/13 with autoit, but I remeber very little. We should think a bit more to find a way to create this link system. 

The first way could be to give to this little program a CLI usage and so we could generate a link to the exe file giving parameters corresponding to the right configuration

PS. you can tag this as feature request

**Comment by snappysnappydog** [comment]
This should be relatively easy to add.

For reference:
[AutoIt Command Line Parameters](https://www.autoitscript.com/autoit3/docs/intro/running.htm#CommandLine)

**Comment by giacomocerquone** [comment]
So as I was writing in the other issue, we need to decide if we eant to break the app in two exes, one for the cli part and one for the GUI that we've already have.

The point is that if we decide to break in two pieces, we can also think to make a full CLI version of the program because in this situation, unless you compile for CUI app https://www.autoitscript.com/forum/topic/103759-how-to-make-console-application/, we can't answer through standard output (the console)

**Comment by KurtisLiggett** [comment]
Since the single-instance issue is a programming feature, we can just program some command line checks in before closing the instance.

Something like this ?
```AutoIt
If $CmdLine[0] > 0 Then
	;check command line parameters...
	if .... Then
		;apply the profile
		;...
		;close the instance
		;...
	EndIf
EndIf

Global $iMsg = _WinAPI_RegisterWindowMessage('newinstance_message')
If _Singleton("Simple IP Config", 1) = 0 Then
	_WinAPI_PostMessage(0xffff, $iMsg, 0x101, 0)
	Exit
EndIf
```

**Comment by giacomocerquone** [comment]
What do you mean by "we can just program some command line checks in before closing the instance"?
To execute the code you posted, unless you compile a new exe, you can't have the app running.

This is what I was writing: a new cli.au3 file that is imported by the "main" and call `Exit` after checking the params, loading the profiles and applying the one choosed.

```autoit
If $CMDLINE[0] Then
  Switch $CMDLINE[1]
      Case 'set-config'
        ;Code for configuration
        _loadProfiles()
        $newfile = ""
        ;profilelist ([profilenumber][fixed=1])[property]
        $newfile &= $propertyFormat[0] & "=" & ($profilelist[2][1])[0]
        MsgBox(0x40000, "asd", $newfile)
      Case Else
          MsgBox(0x40000, 'Incorrect CLI usage', _
                  'Command used:' & @CRLF & $CMDLINERAW & @CRLF & _
                  @CRLF & 'Use "help" for the available commands.')
  EndSwitch
  Exit
EndIf
```

I'm trying to reuse as much code as possible but it's very hard to follow the logic... let me tell you that AutoIt isn't a really good programming language for coding real application.
Your project as an incredible amount of global variables and it's basically based only on side-effect functions, they never return a value.

**Comment by KurtisLiggett** [comment]
I just meant the following piece of code checks if another instance is already running. If it is, then close the new instance.
```AutoIt
If _Singleton("Simple IP Config", 1) = 0 Then
	_WinAPI_PostMessage(0xffff, $iMsg, 0x101, 0)
	Exit
EndIf
```
We just need add your new logic before it to prevent it from being exited too early.
Sorry if you already knew this and I'm just missing your point...

**Comment by KurtisLiggett** [comment]
> I'm trying to reuse as much code as possible but it's very hard to follow the logic... let me tell you that AutoIt isn't a really good programming language for coding real application.
Your project as an incredible amount of global variables and it's basically based only on side-effect functions, they never return a value.

Yes, this started as a fun AutoIt project / proof of concept but it grew into something usable. I will work on commenting and cleaning up the code which should help a lot. MANY of the global variables can be eliminated with better programming (returning values, for example).

I agree that other programming languages may be better in some aspects, but AutoIt has many things going for it.

**Comment by giacomocerquone** [comment]
> We just need add your new logic before it to prevent it from being exited too early.
Sorry if you already knew this and I'm just missing your point...

No, actually I ask you sorry! I didn't understand that code and now I realized that is needed to show a message to the user if there is already an instance of the program running.
So are you sure you don't want to give to the user the possibility to set a configuration through a link when the program is in traybar for example? (As I said this could be done releasing a full cli version of the program that ships with the GUI version)
As soon as we decide this, I'll ask you one thing on the code and then I think I can push a PR 😄

**Comment by giacomocerquone** [comment]
> I agree that other programming languages may be better in some aspects, but AutoIt has many things going for it.

Absolutely True... OOP would be awesome but I must say, procedural programming can give you satisfaction if you avoid certain practices.
Point is that despite being a very powerful language, it isn't considered worthy of consideration in the community and a lot of features are missing like async programming, functional programming, OOP etc.

Anyway switching to "pure return function" should put this project in order, and I'm very happy to help you :)

**Comment by KurtisLiggett** [comment]
> So are you sure you don't want to give to the user the possibility to set a configuration through a link when the program is in traybar for example?

This is what I would like, but I'm open to suggestions.
Case 1:  If the program is running and someone tries to open it again, they get the 'already running' message. But I think we should make this optional through the settings window in case someone wants to run multiple instances for whatever reason.
Case 2:  If it is running and they run it from a link (run with parameters), we allow the 2nd instance to run then popup a tray tooltip when it is done saying 'profile applied successfully' or something like that.

We can make a full cli version, but I would still like to have the above functionality in the GUI version.

**Comment by giacomocerquone** [comment]
> We can make a full cli version, but I would still like to have the above functionality in the GUI version.
This seems fair to me

Sorry but I thought autoit didn't allow opening mulitple instances... if it does, I'd definitely like to put no control. I mean:
Without any setting, a user is always allowed to open multiple instances and if a user click on a link, a new instance get opened and closed

**Comment by KurtisLiggett** [comment]
> if a user click a link and another instance is in the tray we give no warning, just "config applied" and kill the instance opened for the link

I'm good with this

> if a user open a new instance of the program and another one is in the tray, just give a warning

I like this and we can make it the default behavior, but I still would like to have a setting if someone only wants to allow 1 instance. This setting would have nothing to do with the links mentioned above, so the links will allow be allowed to run.

**Comment by giacomocerquone** [comment]
Perfect then, it is working as desired as of now.
Just one thing now: I see the apply function that gets all the information through guictrlread and I obviously can't reuse that function.
Now, what do you say if I recreate an apply_cli function based on the other function in order to make it gets the info directly from the loadprofile function?

**Comment by KurtisLiggett** [comment]
That would be fine.
It'd be better (but more work) to make the _apply function accept adapter and ip info as parameters, so the cli and gui would be calling the same function.

**Comment by KurtisLiggett** [comment]
This feature has been added.
